text/0069-api-v2.md

```diff
@@ -0,0 +1,172 @@
+# TiKV API V2
+
+## Motivation
+
+`API V2` is a set of breaking changes that aim to solve serval issues with current RawKV (hereafter referred to as `API V1`):
+
+1. RawKV is not safe to use along with TxnKV. By solving this, TiDB will be able to support RawKV as the table's storage engine, which will enrich TiDB's use cases.
+2. RawKV TTL is controlled by Store configuration. Switching the configuration will cause data corruption in silence.
+3. RawKV TTL is encoded into the value by appending 8-bytes UNIX timestamp to the end of the value, therefore it's hard to introduce other encode afterward.
+4. RawKV compare_and_swap is easy to make mistake.
+5. It could be nice if we can deploy multiple applications on one TiKV cluster.
+
+## Detailed Design
+
+#### New key-value codec
+
+This RFC introduces a new key encode to RawKV and TxnKV, and a new value encode to RawKV, which will allow the RawKV to be used along with TxnKV and also allow TiKV to flexibly add meta, e.g. the TTL, to a RawKV value.
+
+In addition, keys will be contained in keyspaces, where the keys in different keyspace are totally independant. If keyspace is not specified, the keyspace 'default' will be used.
+
+The `API V2` is enabled by a switch on PD. Since it changes the storage encode, it will be not compatible to switch between `API V1` and `API V2` while there are non-TiDB data in TiKV. TiDB data is specially treated in order to not be affected by the change.
+
+#### Key Encode
+
+Once the `API V2` is enabled, the key will be starting with either:
+
+1. `m` or `t`: TxnKV key. Used by TiDB.
+2. `x{keyspace prefix id}`: TxnKV key.
+3. `r{keyspace prefix id}`: RawKV key.
+4. `a{keyspace prefix id}`: RawKV atomic key, on which you can use `CompareAndSwap`.
+
+The `{keyspace prefix id}` is the [keyspace](https://github.com/tikv/rfcs/pull/39) prefix for seperating keys of different keyspace. It should be an vary-length integer whose highest bit of every byte denotes whether the next byte is still part of the integer. The client will fetch the prefix from PD by the keyspace name specified by the user when initializing the client, so it means that the keyspace prefix is valid during the session, in other words, the change on keyspace on PD will not take effect on running seesions.
+
+In addition, the user key in every mode will be encoded by `Memory Comparable Encoding` as TxnKV does in API V1.
+
+#### RawKV Value Encode
+
+If the key is started with `r` or `a`, then the value can be either:
+
+1. `{0x0}{data}`
+2. `{0x1}{TTL expire timestamp}{data}`
+
+### Keyspace Management
```

Comment:
how about simplifying the HTTP methods to only use GET and PUT, and unifying the API to start with `/keyspace` like the following:

```shell
# list all keyspaces
GET "{pd_address:status}/keyspace/list"

# add keyspaces
curl -X POST -d "{json}" "{pd_address:status}/keyspace/add"

# delete keyspaces
curl -X POST -d "{json}" "{pd_address:status}/keyspace/delete"
```

And we also lack a method to update the name of an existing keyspace, how about adding the following HTTP API:

```shell
# update keyspaces
curl -X POST -d "{json}" "{pd_address:status}/keyspace/update"
```

after adding the update method, the whole CURD operations on keyspace are supported.

Reply:
Not using REST makes things more complicated for me.

This kind of topic is a distraction for this RFC and any other with APIs. Instead we should have a consistent standard for how PD APIs are implemented (for now this would be determined by looking at existing APIs). If we want to change or formalize this standard, there should be a separate RFC.

Reply:
yes. let's focused on the APIs for keyspace management at this RFC.